csrc/optimization/remove_empty.cpp

```diff
@@ -0,0 +1,469 @@
+//! their extents to a constant 0. Here we check for those constant zero
+//! extents.
+bool isTVEmpty(TensorView* tv) {
+  return !emptyAxes(TensorDomain::noReductions(tv->getMaybeRFactorDomain()))
+              .empty();
+}
+
+//! EmptyTensorRemover performs a backward traversal of the Fusion. When it
+//! detects a TensorView that has at least one extent that is zero, we do the
+//! following:
+//!
+//!   1. If the empty Tensorview is a Fusion output, we replace it with a
+//!   TensorView created by `full` having the same shape. Since the original
+//!   tensor is empty, there is nothing to compute, so this eliminates a branch
+//!   of trivial code.
+//!   2. If the empty TensorView is the input of a `cat` op along the empty
+//!   dimension, we replace the cat op with a new one having the empty input
+//!   removed.
+//!   3. If the empty Tensorview is the input to a ReductionOp or WelfordOp and
+//!   the empty dimensions are reduced, we replace the op with `full` since
+//!   there are no elements being reduced. Note that if any empty axes are not
+//!   reduced, we will not encounter this case since it will have been removed
+//!   earlier in the backward traversal under condition 1.
+//!   4. If the empty Tensorview is the input to a PadOp (which is not input to
+//!   a CatOp) then we replace the pad with `full(pad_value)`.
+//!   5. If empty TensorViews are input to an MmaOp and they are empty in
+//!   contracted axes, we replace with `full({m, n}, zeroVal())`.
+//!
+class EmptyTensorRemover : public DeadCodeRemover {
+ public:
+  EmptyTensorRemover(Fusion* fusion) : DeadCodeRemover(fusion) {}
+
+ protected:
+  using DeadCodeRemover::handle;
+
+  //! If tv is a fusion output, we check whether it is empty and if so, replace
+  //! it with full(). For non-outputs that are not inputs, we simply check that
+  //! the tensor is not provably empty.
+  void handle(TensorView* tv) final {
+    if (tv->isFusionInput()) {
+      // Skip inputs since they do not have a definition to redefine
+      return;
+    }
+
+    if (tv->isFusionOutput()) {
+      const auto rfactor =
+          TensorDomain::noReductions(tv->getMaybeRFactorDomain());
+      const auto empty_axes = emptyAxes(rfactor);
+      if (!empty_axes.empty()) {
+        std::vector<Val*> shape(rfactor.size());
+        std::transform(
+            rfactor.begin(), rfactor.end(), shape.begin(), [](IterDomain* id) {
+              return id->extent();
+            });
+        for (auto ax : empty_axes) {
+          shape[ax] = fusion()->zeroVal();
+        }
+        auto dtype = tv->getDataType().value();
+        auto new_tv = full(shape, fusion()->zeroVal(dtype), dtype);
+        replaceTV(tv, new_tv);
+      }
+    } else if (allUsesDead(tv)) {
+      // TensorViews that are not Fusion inputs or outputs and which have no
+      // uses are dead, so remove them and skip processing them and their
+      // definition.
+      markDeadAndMaybeRemove(tv);
```

Comment:
I pushed an update to the API. You're right it was confusing, so hopefully this will help. Now, child classes should only modify the Fusion by calling `replaceVal` and `removeVal`. Those will mark the `Val` as dead, and will remove as many dead `Val`s and `Expr`s as possible. Each returns a `bool` indicating whether it will actually be removed from the `Fusion`. The removal actually occurs after the traversal, so it is safe to dereference pointers at any point during traversal.